Restore comparingsearch level

mickaelistria opened the pull request.

snjeza committed:
Using javac ASTParser outside of OSGi environment

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

mickaelistria committed:
Cache and reuse symbols

Instead of always re-reading the jars for symbols, we create a local
cache that is capable of recreating the symbol into any context reliably
enough from memory.
This should drastically improve performance.

mickaelistria committed:
Fixes symbol cache

* ForAll/generic methods

Commit:
Fixes testMethodDeclarationsElementPatternMultipleParamArguments02

Signed-off-by: Rob Stryker <stryker@redhat.com>

snjeza committed:
Fix ASTConverterTestAST8_2.test0602

Commit:
Fix regressions

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes testMethodReferencesStringPatternSingleParamArguments09 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Don't suggest keywords in an array initializer

eg. there should be no keywords when completing at the `|`

```java
int[] a = new int[] { | };
```

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOMSearch] - Fixes testMethodReferencesStringPatternSingleParamArguments02

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fall back to search to find `IType` for a `JavacTypeBinding`

Fixes part of #1401

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Small fixes for record bindings

mickaelistria committed:
Map more problems

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1394
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1395

mickaelistria committed:
Update GitHub action

mickaelistria committed:
Add workaround repo for repo.eclipse.org outage

Commit:
[DOMSearch] Fixes testMethodReferencesStringPatternSingleParamArguments06

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Try archiving JUnit reports in GitHub actions

mickaelistria committed:
More GH JUnit reporting

Commit:
Fixes testMethodReferencesElementPatternSingleTypeParameter04 and one other

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[JavacConverter] Imports are having static set twice, which breaks them

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
NPE Check

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOMSearch] - Fix testMethodDeclarationsElementPatternMultipleParamArguments05 and one other

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
NPE

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
More fixes to CachingClassSymbolClassReader

* Synchronize init of reflection
* Do not share the signatureBuffer, keep it instance specific to avoid
concurrent modification. Sharing doesn't bring any benefit anyway.

Commit:
[DOM Search] More method searches

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Only complete `default` inside annotations

mickaelistria committed:
Replace name instead of inserting when possible

mickaelistria committed:
Make ASTConverterMarkdownText more tolerant

A parser may decide to keep a TextElement as " blah" with (offset,
length)=(N,5); or as "blah" with (offset,length)=(N+1,4). Both are
valid.
Make test tolerate both.

mickaelistria committed:
Fix completion for `blah.|  unrelatedMethod()`

Commit:
[DOM Search] Fixes 6 in JavaSearchGenericMethodExactTests

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Search skips visiting EnumConstantDeclaration name

The parent EnumConstantDeclaration is already checked, no need to
re-check the name.

Commit:
Fix / Swallow 20 IllegalArgumentExceptions when comparing unresolved patterns

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Match MethodReference (by name)

Flag needs to resolve bindings to evaluate whether match is real or not.

Commit:
[DOM Search] Bandaid Fix for problem with wildcard key being completely wrong
  Fixes 15

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Fix 12 - fixes isQuestionMark and ensures wildcard keys are parseable

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Match methodReferences

datho7561 committed:
Do not suggest keywords in empty block

I disagree personally but this is what JDT does.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Do not filter out types from file when completing field type

eg.

```java
public class MyLinkedList {

  public MyLinkedList(MyLinkedList next) {
    // ... assign next to a field
  }

  MyLink|
}
```

We were filtering `MyLinkedList` out because it was "default" completion,
but we were disabling default completion in this case.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix some regressions

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
NPE

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix test0264 in ASTConverter15JLS4Test

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Random Fixes in AST Converter tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix JavacAnnotationBinding.getAllMemberValuePairs()

also add default values.

Commit:
[DOM Search] Fixes testParameterizedTypeMultipleArguments01 in equivalent tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Cleanup an unnecessary hard-coded signature

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Fixes testParameterizedType_Bug83713

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes testStringAnyStrings06 - name match with wildcard

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Improve JavacMethodBinding.isRaw()

Use info from methodType

mickaelistria committed:
[Search] Fix qualifiedSourceName

qualifiedSourceName is the qualified name (including parent types for
nested/local types) but without including the package.

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1430

mickaelistria committed:
[Search] Compare simpleName when no qualifier is used

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1430

mickaelistria committed:
Revert "[Search] Fix qualifiedSourceName"

This reverts commit c02fecb00640ad097d55ec4974804cee5678d071.

Commit:
Fix an NPE in testMethodReferencesElementPatternMultipleTypeParameter03

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes testParameterizedArrayTypeSingleArgument04

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes testStringNestedSingleArgument04

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Only add unresolved matches when there's some part that matches

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Allow parse() to return result even if some units are missing

+ prevent a ClassCastException

mickaelistria committed:
Implement getDOMASTMethodBinding

mickaelistria committed:
Support for ConstructorInvocation in search

mickaelistria committed:
Fix type comparison in search

Keep arrays

mickaelistria committed:
Fix for field references in field initializers

* Don't skip them
* Fix resolving the enclosing Java element

mickaelistria committed:
[Search] Support OrPattern for DOM

Commit:
Fix test testMethodReferencesElementPatternMultipleTypeParameter02 - properly prefer weaker claim

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix an NPE in DOMMethodLocator

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix an IllegalArgumentException when calling member.getType()

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
[Search] Do not visit AnnotationTypeDeclaration name

Commit:
Comment out 3 tests that keep killing my env

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fix some constructor completion tests

- Fix constructor completion test that were failing due to the wrong
  signature
- Complete first parameter for constructors

When the cursor is here (at the `|`):

```java
MyClass obj = new MyClass(|)
```

Then the appropriate parameter should be suggested (in addition to the
constructor itself, as well as an anonymous constructor).

When the cursor is here:

```java
new MyClass().|
```

It should suggest appropriate fields and methods.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Mostly fix testSearchFieldInBinaryNoResolution - searches for field patterns should not return matches with non-field java elements

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
[Search] Honor `searchFor`

mickaelistria committed:
Revert "Mostly fix testSearchFieldInBinaryNoResolution - searches for field p…"

This reverts commit 57a461f8ad6080c8000b07e98a2abb5adb09008f.

mickaelistria committed:
fix search for binary types

* Compare JavaElements directly
* Improve getJavaElement() for JavacTypeBinding to skip working copy
* Fix 1 test

mickaelistria committed:
[Search] Support reference to annotation member

Commit:
Fix various NPEs

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Better support for searching MemberValuePair

mickaelistria committed:
[Search] Improve comparison of method return type

mickaelistria committed:
Improve getting level for arrays

mickaelistria committed:
[Search] Improve super type reference

Do not rely on flavors (that are set by the ECJ parser, not being used
here)

mickaelistria committed:
[Search] Fix error when matching package declaration

mickaelistria committed:
Fix resolving JavaElement for binary types/methods

* Special rules for when using a mock name for Javac (this could
probably be improved by getting ASTParser to pass through the actual
class name and resusing it)
* Use resolved types signatures from binding to retrieve method
reference
* Better signature for parameterized/generic types

mickaelistria committed:
Match SingleMemberAnnotation value()

mickaelistria committed:
[Search] Improve filters when matching super-type

mickaelistria committed:
[Search] add support for StrigLiteral in SingleMemberAnnotation

mickaelistria committed:
[Search] Match type parameters from type declaration

+ fix matching wildcards

mickaelistria committed:
[Search] prefer returning actual member method (instead of similar one)

mickaelistria committed:
[Search] Add localElement to TypeParam ref match

mickaelistria committed:
Fix JavacTypeBinding.getErasure() for arrays

Commit:
Fix testTypeNestedSingleArgument03 and one or two others while preventing regressions

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix JavacMethodBinding for varargs

JDT does not care about the particular method invocation parameters
type, but more about the method declaration.

mickaelistria committed:
Keep track of the .jar when processing binary type

Commit:
Fix testBug549413_017 - use correct working copy resolver, and more

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Find SUPER_INVOCATION and flag them as expected

mickaelistria committed:
Match inside binary files

datho7561 committed:
Implement completion for modules in Javadoc

eg. completion at the `|`

```java
public class Main {

  /**
   * @see java.base/|
   */
  public static void myMethod() {
  }

}
```

Also this case:

```java
public class Main {

  /**
   * @see java.base/java.lang.|
   */
  public static void myMethod() {
  }

}
```

Also this case:

```java
public class Main {

  /**
   * @see java.base/java.lang.String#ha|
   */
  public static void myMethod() {
  }

}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Address bug in javadoc type completion

It shouldn't fully qualify the type if the type is imported

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Only suggest types in the `@throws` Javadoc

Should fix 3 tests

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fixes testMethodReferencesElementPatternMultipleParamArguments01 - dom search with type args for receiver of method call

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix searching java element in initializer

The domOccurrence value must be computed for the particular type, not
relative to all members

datho7561 committed:
Javadoc tag completion for unparented comments

Should fix 2 tests

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve completion for method references in Javadoc

Specifically, only types (and primitive types) should be suggested at the `|`:

```java
/**
 * @see #myMethod(|)
 */
public class Main {
  void myMethod(String pants) {}
}
```

As well, I fixed the replace range in this case, where it used to be
replacing the method name as well:

```java
/**
 * @see #myMethod(Str|)
 */
public class Main {
  void myMethod(String pants) {}
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix NPEs in package Javadoc completion

package declarations have no parent type declaration,
this was causing some NPEs

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix context for Javadoc tag elements

The "@" should be included in the token.
I needed to adjust how the completions are calculated in order to take
this into account.

Should fix 8 test cases

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix a completion case for `@link`

Fix the following completion case. Completion at the `|` should suggest
`HelloWorld`.

```java
public class HelloWorld {
  /**
   * @link |
   */
  public static class InnerClass {

  }
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Complete types in javadoc text

eg.

```java
public class Outer {
  /**
   * Another cool class is Out|
   */
  public static class Inner {

  }
}
```

should suggest `Outer` and `{@link Outer}` as completions.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Some more unparented comment completion fixes

- Improve selection algorithm of unparented comments in `DOMCompletionContext`
- When completing a type in a Javadoc, do not fully qualify types from
  `java.lang`

Fixes 8

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Some more tag completion fixes

- Handle completion for unparsed inline tags
  - Fixes 1 test case
- Suggest method javadoc tags when an unparented comment is in a type
  declaration
  - eg. suggest `@serialData` for a comment at the bottom of a class
    declaration body
  - I needed to use a hacky method to do this, since I can't just check the
    javadoc's parent nodes (since it's unparented).
  - Fixes 1 test case

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Refining method matching, converting resolveLevel(MessageSend)

mickaelistria committed:
Refining match method declaration when overriding generic super

reintroduce similar checks has MethodLocator.newDeclarationMatch(...).

datho7561 committed:
Refactor "find parent type binding"

Make an instance variable for storing `DOMCompletionContext`.
Then, provide a method for lazy loading the binding of the current type in `DOMCompletionContext`.

As well:
- Delete some unused variables
- Fix the broken accessibility check for IType

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Move true/false binding finder into utils class

The true/false binding finder doesn't rely on much completion context
and can be extracted into a static helper method.

- Move DOMCompletionUtil to DOMCompletionUtil__s__
  - This was a joke naming because there was initially only one method
    in the class, I should have just stuck with the convential name to
    begin with.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Refactor DOMCompletionEngine to use completeASTNodeName(...) methods

- Use switch statement instead of if statements in DOMCompletionEngine

Fixes #1501

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve SwitchCase completion, add break & continue completion

- For break and continue, suggest applicable statement labels
- For SwitchCase, handle completing enums when the token is empty
- Also, handle some SimpleName cases by using the parent as the context
  instead of handling them in `completeSimpleName`.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Properly set bindingIsDeclaration

mickaelistria committed:
Do not take return type into account when matching methods

When matching known methods (which have a "declaringSimpleName", then we
don't need to care about the return type to match the overridden
methods.

mickaelistria committed:
Support for matching MethodRef

mickaelistria committed:
Match reference to implicit super-constructor

datho7561 committed:
Expected types for yield statement

This makes the result consistent with what's done with `return`,
but inconsistent with what's done upstream.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
More support for SearchMatch.localElement

+ fix for VariableBining.getDeclaringClass()

datho7561 committed:
Respect ignored completion proposals

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix Javadoc tag replace range

Fixes 8.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Match recovered types

mickaelistria committed:
Match recovered types (2)

mickaelistria committed:
Support read/write access info on variable/field match

mickaelistria committed:
Fix computation of mustResolve

particular locators can force resolution, read the mustResolve state
after processing the locators

+ make any string match null/ignored name in pattern

mickaelistria committed:
Support otherElements in variable/field type declaration match

robstryker committed:
[DOM Search] modify tests to not compare match level for specific situations

Signed-off-by: Rob Stryker <rob@oxbeef.net>

robstryker committed:
[DOM Search] Fix some NPEs

Signed-off-by: Rob Stryker <rob@oxbeef.net>

datho7561 committed:
Replace range for qualified type in a Javadoc

Signed-off-by: David Thompson <davthomp@redhat.com>

robstryker committed:
Fix testStringNestedSingleArgument03

Signed-off-by: Rob Stryker <rob@oxbeef.net>

robstryker committed:
Fix testMethodDeclarationsStringPatternMultipleParamArguments04

Signed-off-by: Rob Stryker <rob@oxbeef.net>

robstryker committed:
Fix regressions

Signed-off-by: Rob Stryker <rob@oxbeef.net>

datho7561 committed:
Improve completion of nested types in Javadoc

This case now works the same as it does upstream:

```java
package dev.datho7561.utils;
/**
 * Outer.Inn|
 */
public class Main {

}
class Outer {
  class Inner {
  }
}
```

There is a bug in jdt.ui which makes it not work in practice though.
It tries to import `Inner`, despite the fact that it's in the same
package and no "additional completion" is provided.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion of variable before `this`

- [x] TODO: test case

Fixes #1433

Signed-off-by: David Thompson <davthomp@redhat.com>

robstryker committed:
Fix 6 issues in DOMMethodLocator

Signed-off-by: Rob Stryker <rob@oxbeef.net>

Fix an NPE

Signed-off-by: Rob Stryker <rob@oxbeef.net>

mickaelistria committed:
Fix missing new line in p2.inf

mickaelistria committed:
Multi variable declarations in same switch statement

mickaelistria committed:
Fix qualifiedName for local types

mickaelistria committed:
setEscapedValue for CharacterLiteral

mickaelistria committed:
Return null declaring method for fields

mickaelistria committed:
Properly set binding flag on AST

mickaelistria committed:
Match alternative message from Javac for missing method

mickaelistria committed:
Fix potential NPE in DOMMethodLocator.resolveLevel()

mickaelistria committed:
Filter field/method matches according to fineGrain

mickaelistria committed:
Disable irrelevant check failing because of serialVersionUUID

mickaelistria committed:
Support searching primitive types

mickaelistria committed:
Partial support for fineGrain references

mickaelistria committed:
Add missing visited types for search

mickaelistria committed:
skip type params for fineGrain patterns

robstryker committed:
Comment out some debugging output

Signed-off-by: Rob Stryker <rob@oxbeef.net>

datho7561 committed:
Fix bad method signature

Uses a regex to strip some extra information that's present in keys but
not signatures.

Includes a new test case to cover this behaviour, since there is no
existing test for this.

Includes an unrelated fix (relevance number when a raw type is expected) to get the new test case passing.

Fixes #1414

Signed-off-by: David Thompson <davthomp@redhat.com>

rgrunber committed:
Update versions in parent POM references for javac bundle

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>

datho7561 committed:
Completion for @docroot inline tag

Signed-off-by: David Thompson <davthomp@redhat.com>

robstryker committed:
Fix testMethodReferencesElementPatternMultipleParamArguments02 and others
More work at unifying logic for resolveLevelForNodeWithMethodBinding

Signed-off-by: Rob Stryker <rob@oxbeef.net>

datho7561 committed:
Fix relevance for qualified package completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve relevance calculation for NON_STATIC

Fixes 2 Javadoc-related test cases

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Set raw on matches

mickaelistria committed:
Restore comparing search level

robstryker commented:
I'd like to see failures down a bit more before we go doing this, but it's ultimately up to you.

mickaelistria commented:
I agree, I was expecting a bit less than a hundred failures, but it's much more. I am targetting a merge when the extra failures are up by 100 or less.

mickaelistria committed:
Treat raw method calls as ERASURE_MATCH w/ pattern has type params

mickaelistria committed:
Filter reported matches

+ improve rule for generic methods

mickaelistria committed:
Matching raw methods as EQUIVALENT

robstryker committed:
[DOMSearch] Fix some NPE's where we have null elements in the array of other java elements in type matches

Signed-off-by: Rob Stryker <rob@oxbeef.net>

mickaelistria committed:
Port some updateMatch methods

mickaelistria committed:
Improve updateMatch

Better resolve pattern binding

datho7561 committed:
Fixes to `super` completion

Should fix 2

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix resolveConstructorBinding

Use constructorType, which keeps ref of type parameters

mickaelistria committed:
Better findType in DOMPatternLocator

mickaelistria committed:
Convert methodRef with annotated types

mickaelistria committed:
Search visit TypeMethodReference

mickaelistria committed:
search visit SimpleType.annotations()

mickaelistria committed:
Fix possible issue with erroneous constructorType

mickaelistria committed:
Fix case of no typeArgs

mickaelistria committed:
Don't match filter enumConstants for generics

mickaelistria committed:
Fix match variables from children lists

mickaelistria committed:
match resolve modules

mickaelistria committed:
Fix arrayType dimension annotation conversion

mickaelistria committed:
Match methodReference even without fineGrain

mickaelistria committed:
Fix package declarations in test

mickaelistria committed:
Improve matching/reporting method references

mickaelistria committed:
Find binding when matching LambdaExpression

mickaelistria committed:
Use Signature in LambdaMethod model element

mickaelistria committed:
Improve matching LambdaExpression

mickaelistria committed:
Fix conversion of dimension

`String tab @Annot[] ` was converted to `String @Annot tab @Annot []`

mickaelistria committed:
More matches

mickaelistria committed:
Fix getJavaElement for lambda parameter binding

datho7561 committed:
Handle suggesting overlapping methods for parent classes

eg. both `equals` and `Hello.this.equals` should be suggested here:

```java
public class Hello {
  void myMethod() {
    Object obj = new Object() {
      void otherMethod() {
        equa|
      }
    }
  }
}
```

Note that in practice this completion doesn't work
(likely due to a bug in jdt.ui),
but it should fix some test cases.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix tests: `pacakge` -> `package`

mickaelistria committed:
Properly resolve Intersection type in search

mickaelistria committed:
Serch match CreationReference

mickaelistria committed:
Nested lambda bindings

mickaelistria committed:
Fix lambda binding signature

mickaelistria committed:
Match/resolve types declared in lambda

mickaelistria committed:
AndLocator

mickaelistria committed:
Improve fineGrain match for methodReferences

mickaelistria committed:
Better match intersection/unions

mickaelistria committed:
Search recovering statements

mickaelistria committed:
Fix matching references to nested types

datho7561 committed:
Improve import completion based on some failing tests

1. Improve import completion for simple name

eg.

```java
import pack|
```

This case should suggest subpackages in `pack` but also `pack1`.

2. Do not suggest multiple of the same type when completing an import

We aren't checking for duplicate type names when completing types in a
package.

Fixes #1596

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Match some modules

mickaelistria committed:
Properly build ModuleReferenceMatch

datho7561 committed:
Improve catch clause type completion

Handle this case:

```java
try {
  // ...
} catch (|) {

}
```

Gets 6 test cases passing

This PR also handles collecting top level types when running
`scrapeAccessibleBindings()` and completing types for method parameters.

- Add test for completing method parameters
- Fix test for yield completion (might as well)

Fixes #1597

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix qualified type completion in catch clause

We weren't properly excluding the already caught exceptions

Fixes #1599

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Do not suggest interface,enum,annotation as caught type

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Update README.md

mickaelistria committed:
Imprive Zip cache so closed FS are not returned

mickaelistria committed:
Better search unresolved module

mickaelistria committed:
Improve setting modulePath

datho7561 committed:
Improve "uninteresting" relevance for catch clause

Fixes #1607

Fixes 5 cases

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix JavacModuleBinding.getJavaElement()

mickaelistria committed:
Return packageBinding even for empty packages

mickaelistria committed:
Simplify/fix package reference locator

mickaelistria committed:
Exclude package declaration and treat no binding as package

mickaelistria committed:
Return type when matching import declaration in binary

mickaelistria committed:
Fix module name match

Invert operands to better match prefix/regexp

robstryker committed:
Fix testStringAnyStrings01 and 3 others

Signed-off-by: Rob Stryker <rob@oxbeef.net>

robstryker committed:
Missed a line deletion

Signed-off-by: Rob Stryker <rob@oxbeef.net>

robstryker committed:
Fix testTypeReference35 and others

Signed-off-by: Rob Stryker <rob@oxbeef.net>

mickaelistria committed:
Better set module-source-path

mickaelistria committed:
Consider module filtering when matching type reference

mickaelistria committed:
Match module without resolving graph if possible

mickaelistria committed:
Ignore modules when reading module-info from jar

mickaelistria committed:
Do not match method name as field

mickaelistria committed:
Do not put corrupt jar in class/module path

mickaelistria committed:
Better match level for method declarations with generics

mickaelistria committed:
Fix fine grain matching permitted types reference

datho7561 committed:
Fix completion and relevance for super field

- Do not suggest methods from enclosing types after `super.|`
- These completions should count as NON_STATIC
- These completion should not count as either QUALIFIED or UNQUALIFIED

Fixes 1 case

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix `CompletionTests.testBug385858c` (bad context expected type signature)

- Fix calculation of signature of expected types in completion context.
  It wasn't handling inner types properly.
  - use our method from `SignatureUtils` to handle this
- Handle "start of constructor" token location calculation (also related
  to completion context)

Fixes 3

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Assorted completion bug fixes

- Don't complete enum constant declarations
- Better handling of completion of qualified `super` and `this`
  (take into account if you're in an interface)
- Improve `JavacConverter` to handle "field access" split across two lines
  as in `CompletionTests.testBug573632`
  (signatures don't match properly so it's not passing)

Should fix 4 and improve many more

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Match import module

mickaelistria committed:
Fix read/write match check for fields

datho7561 committed:
Fix for `CompletionTests.testBug574982`

Another portion of the fix depended on addressing https://github.com/eclipse-jdt/eclipse.jdt.core/issues/4247,
which is now resolved

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improve naming of java files from jar

datho7561 committed:
Suggest package name for all inner classes outside current one

Fixes 1 test case

Signed-off-by: David Thompson <davthomp@redhat.com>

robstryker committed:
Fixes JavaSearchMultipleProjectsTests.testPackageReference1

Signed-off-by: Rob Stryker <rob@oxbeef.net>

datho7561 committed:
Display type arguments of method parameters in method completion

Fixes #1629

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improve matching/filtering anonymous types

datho7561 committed:
Protected members from class in same package should be suggested

Fixes #1630 , fixes 1 test case

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Do not match methodInvocations without bindings

mickaelistria committed:
Refine setting local element

Also fix DOMVariableBinding.isParameter() for caught Exception

mickaelistria committed:
Avoid exception converting AST with empty variable fragments

mickaelistria committed:
Fix matched IJavaElement for record component

mickaelistria committed:
Better detect record components bindings

datho7561 committed:
Don't convert module when using Java 8

There are other issues that arise when having a `module-info.java` in a Java 8
project, but this makes sure that we at least don't try to convert the AST nodes,
which prevents an exception.

Fixes #193

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix completion of variables from an instanceof expression on same line

eg. after `|`

```java
public class CompleteInInstanceof {

	public void myMethod() {
		Object o = null;
		boolean a = o instanceof RequiresDirective requires && useRequiresDirective(req|);
	}

	static class RequiresDirective {
		private int a;
		public RequiresDirective(int a) { this.a = a; }
		int getA() { return a; }
	}

	public static boolean useRequiresDirective(RequiresDirective requiresDirective) {
		return requiresDirective.getA()  > 5;
	}
}
```

Fixes #1593

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Better match reference to record components

Record components are often modeled as ILocalVariable, which might use
different parent in a case or another. Instead of strictly checking for
equal model element, we check that they are equivalent (name and
position)

mickaelistria committed:
Fix package in tests

mickaelistria committed:
Do not match package binding as type reference

mickaelistria committed:
Do not set module source path is module-info.java doesn't exist

mickaelistria committed:
Fix ranges for module-qualified member reference

mickaelistria committed:
Don't match RecordDeclaration name

mickaelistria committed:
Ignore completion tests for the moment

Completion tests are currently ignored as functional coverage for
completion is already good and many tests are really failing for
insignificant details (such as relevance which is overridden in UI
anyway). So we skip the test to avoid noise and let dogfooding and
actual usage feedback drive further improvements to completion.
We may re-enable those tests later.

mickaelistria committed:
Process possibleMatch.getSimilarMatch() for binary

mickaelistria commented:
Adds ~200 failures, let's wait a bit before merging it.
